src/hotspot/share/interpreter/interpreterRuntime.cpp

```diff
@@ -733,20 +733,16 @@ JRT_ENTRY_NO_ASYNC(void, InterpreterRuntime::monitorenter(JavaThread* thread, Ba
 JRT_END
 
 
-//%note monitor_1
-JRT_ENTRY_NO_ASYNC(void, InterpreterRuntime::monitorexit(JavaThread* thread, BasicObjectLock* elem))
+JRT_LEAF(void, InterpreterRuntime::monitorexit(JavaThread* thread, BasicObjectLock* elem))
 #ifdef ASSERT
   thread->last_frame().interpreter_frame_verify_monitor(elem);
 #endif
-  Handle h_obj(thread, elem->obj());
-  assert(Universe::heap()->is_in_or_null(h_obj()),
-         "must be NULL or an object");
-  if (elem == NULL || h_obj()->is_unlocked()) {
-    THROW(vmSymbols::java_lang_IllegalMonitorStateException());
-  }
```

Comment:
In the case of an unbalanced monitorexit(), you are losing
the throwing of the IllegalMonitorStateException here.
At one point, we had at least one test that used JNI MonitorExit()
to induce an unbalanced monitorexit() and it verified that the
IllegalMonitorStateException was thrown.

Reply:
But JNI MonitorExit() calls ObjectSynchronizer::jni_exit() instead and that calls check_owner() which will throw java_lang_IllegalMonitorStateException() if the thread is not the owner. I agree we might still need to throw java_lang_IllegalMonitorStateException() here for the release bits though. Maybe we still need to cover the case of the classfile having wrong bytecodes and there is an extra monitorexit?

Reply:
The JNI MonitorExit() call will succeed because the calling thread is
the owner of the monitor. The subsequent monitorexit byte code
should fail because the calling thread is no longer the owner of
the monitor. It's a way of simulating an errant situation.

Reply:
Ok, so the scenario would be:
1) Thread calls monitorenter bytecode
2) Thread calls jni_enter()*n and jni_exit()*(n+1) (extra unlock). Also with JNI we always inflate the lock.
3a) If no deflation happens and the thread calls monitorexit bytecode then we hit the assert in https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/objectMonitor.cpp#L1047
3b) If the monitor was deflated then on monitorexit is_unlocked() returns true and we throw java_lang_IllegalMonitorStateException. (We would change this case then).

Is that correct?

Reply:
No jni_enter(). Just jni_exit(). We are trying to provoke an unbalanced
monitorexit() here.

Reply:
Yes, it looks like you can break this with JNI code with and extra jni_exit() for the object, and then call the monitorexit from the interpreter.  This interpreter change is to be consistent with the compiled code that calls via JRT_LEAF, so compiled code would have the same problem.
I can add a fatal to this like:
fatal(obj->is_locked(), "Native code may have unlocked this object");
fatal() would assert in product mode, and errant native code can expect crashes even in product mode.
elem would never be null since it's allocated in the caller.

Reply:
I think you mean assert(obj->is_locked(), "this object must be locked") since fatal() doesn't take a bool...

Reply:
> Maybe we still need to cover the case of the classfile having wrong bytecodes and there is an extra monitorexit?
The checks before the calls to unlock_object() handle this case.

Reply:
Actually, I think it has to be:

Reply:
Update: looks like oop has both is_locked() and is_unlocked()...

Reply:
What happened here? These comments are now completely out of context relative to the code above them? Was this PR rebased?